contracts/secretswap_router/src/contract.rs

```diff
@@ -0,0 +1,378 @@
+use cosmwasm_std::{
+    debug_print, from_binary, to_binary, Api, Binary, Coin, CosmosMsg, Env, Extern, HandleResponse,
+    InitResponse, Querier, StdError, StdResult, Storage, WasmMsg,
+};
+use secret_toolkit::snip20;
+use secretswap::{Asset, AssetInfo};
+use HandleMsg::RecoverFunds;
+
+use crate::{
+    msg::{HandleMsg, Hop, InitMsg, NativeSwap, QueryMsg, Route, Snip20Swap},
+    state::{
+        delete_route_state, read_owner, read_route_state, read_tokens, store_owner,
+        store_route_state, store_tokens, RouteState,
+    },
+};
+
+pub fn init<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    _msg: InitMsg,
+) -> StdResult<InitResponse> {
+    store_owner(&mut deps.storage, &env.message.sender)?;
+    store_tokens(&mut deps.storage, &vec![])?;
+
+    Ok(InitResponse::default())
+}
+
+pub fn handle<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse> {
+    match msg {
+        HandleMsg::Receive {
+            from: _,
+            msg: Some(msg),
+            amount,
+        } => {
+            // This is the first msg from the user, with the entire route details
+            // 1. save the remaining route to state (e.g. if the route is X/Y -> Y/Z -> Z->W then save Y/Z -> Z/W to state)
+            // 2. send `amount` X to pair X/Y
+            // 3. call FinalizeRoute to make sure everything went ok, otherwise revert the tx
+
+            let Route { hops, to } = from_binary(&msg)?;
+
+            if hops.len() < 2 {
+                return Err(StdError::generic_err("route must be at least 2 hops"));
+            }
+
+            for i in 1..(hops.len() - 1) {
+                if hops[i].from_token.address == None
+                    || hops[i].from_token.code_hash == None
+                    || hops[i].from_token.native_denom != None
+                {
+                    return Err(StdError::generic_err(
+                        "cannot route via uscrt. uscrt can only be route input token or output token.",
+                    ));
+                }
+            }
+
+            let first_hop: Hop = hops[0].clone();
+
+            let received_from_token_of_first_hop =
+                Some(env.message.sender) == first_hop.from_token.address;
+
+            let mut received_scrt = false;
+            let mut scrt_is_first_hop = false;
+            if env.message.sent_funds.len() == 1 {
+                received_scrt = env.message.sent_funds[0].amount == amount
+                    && env.message.sent_funds[0].denom == "uscrt";
+                scrt_is_first_hop = first_hop.from_token.native_denom == Some("uscrt".into());
```

Comment:
Doesn't matter if it's "uscrt" or something else. Create an IsNative method on Token and use that for this condition